#don't use modprobe -k
#Sven Ulland <svensven@gmail.com>
diff --git a/keepalived/check/ipvswrapper.c b/keepalived/check/ipvswrapper.c
index ad79f20..497d850 100644
--- a/keepalived/check/ipvswrapper.c
+++ b/keepalived/check/ipvswrapper.c
@@ -687,7 +687,7 @@ string_to_number(const char *s, int min, int max)
 static int
 modprobe_ipvs(void)
 {
-	char *argv[] = { "/sbin/modprobe", "-s", "-k", "--", "ip_vs", NULL };
+	char *argv[] = { "/sbin/modprobe", "-s", "--", "ip_vs", NULL };
 	int child;
 	int status;
 	int rc;
