diff -Nur keepalived-1.1.15/keepalived/vrrp/vrrp_arp.c keepalived-1.1.15.new/keepalived/vrrp/vrrp_arp.c
--- keepalived-1.1.15/keepalived/vrrp/vrrp_arp.c	2007-09-14 17:20:47.000000000 +0200
+++ keepalived-1.1.15.new/keepalived/vrrp/vrrp_arp.c	2007-12-18 19:06:42.000000000 +0100
@@ -22,6 +22,9 @@
  * Copyright (C) 2001-2007 Alexandre Cassen, <acassen@freebox.fr>
  */
 
+/* Work around incompatibility between linux/types.h and sys/types.h */
+#define __KERNEL_STRICT_NAMES
+
 /* local includes */
 #include "vrrp_arp.h"
 #include "memory.h"
